Implementation of map tile cache

n76 opened the pull request:
See: https://github.com/zdila/gpx-animator/issues/77

Implement a map tile cache to avoid downloading the same files each time
a slightly different rendering is tried. User specifies location of cache
and lifetime of cached tiles. If no cache is specified then revert to old
behavior of downloading tiles each time renderer is run.

This is a partial implementation because only the CLI has been updated to
allow specification of cache parameters. The GUI will need updating to
complete this enhancement.

Signed-off-by: Tod Fitch <tod@fitchfamily.org>

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+
+public class TileCache {
+
+    public static BufferedImage getTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
```
As in the other classes, I would like to see method parameters be marked as `final`. Same with the other methods in this class.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+    public static BufferedImage getTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
+
+        if (cachePath == null) {
+            //System.out.println("No cache path defined, downloading " + url);
```
Please remove this commented code. Same with line 55.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+    
+    private static BufferedImage cachedGetTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
+        BufferedImage mapTile;
+        String fname = hashName(url);
```
I guess the `f` stands for `file`? I would like to rename this variable to `filename`. That removes future misunderstanding of other people working on this code.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+        BufferedImage mapTile;
+        String fname = hashName(url);
+        String path = cachePath + File.separator + fname;
+        File f = new File(path);
```
I don't like one-letter-names for variables... :-)

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+            Date fileDate = new Date(f.lastModified());
+            long msBetweenDates = new Date().getTime() - fileDate.getTime();
+            if ((msBetweenDates/1000) > cacheTimeLimit) {
+                System.out.println("Removing cache entry for " + url); 
```
The other two "debug" messages are commented out. Should this be removed, too? Same with line 95.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+            if ((msBetweenDates/1000) > cacheTimeLimit) {
+                System.out.println("Removing cache entry for " + url); 
+                if (!f.delete()) {
+                    throw new UserException("failed to delete cache entry for " + url);
```
Maybe it is a good idea to add the name of the cache file to the error message because it is different from the URL. With the file name, the user could then check the cache directory and the file for problems.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+        if (!f.isFile()) {
+            System.out.println("Cache miss for:" + path);
+            try {
+                FileUtils.copyURLToFile(new URL(url), f, 5000, 10000);
```
How does the HTTP request header look like? In the past when requesting a tile we send "GPX Animator" as the user agent (see line 62). Some APIs require a valid user agent. Otherwise, they might block the requests.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+        }
+        
+        // At this point a current map tile should be in our cache, so
+        // load it for the 
```
load it for the ??? :-)

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+    
+    private static String hashName(String url) throws UserException {
+        try {
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
```
Some background maps need a few hundred tiles. You should initialize the message digest on the object level, not method level. You need this always, there is no use case, where this code is never executed.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
+            return bytesToHex(digest.digest(url.getBytes(StandardCharsets.UTF_8)));
+        } catch (final NoSuchAlgorithmException e) {
+            throw new UserException("error creating hash name " + url, e);
```
Not sure, if I like this. There could be an automatic fallback to the no-cache code. If the cache is not working, that should not prevent the application from generating the video.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/TileCache.java:
```diff
@@ -0,0 +1,135 @@
+        StringBuffer hexString = new StringBuffer();
+        for (int i = 0; i < hash.length; i++) {
+        String hex = Integer.toHexString(0xff & hash[i]);
+        if(hex.length() == 1) hexString.append('0');
```
Please correct the indentation. I had to look several times at the code to identify correctly, that line 131 is not executed based on the `if` on line 130...

McPringle commented on src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java:
```diff
@@ -89,52 +91,52 @@ public GeneralSettingsPanel() {
+		add(outputFileSelector, gbc_outputFileSelector);
 
-				outputFileSelector.addPropertyChangeListener("filename", propertyChangeListener);
+		outputFileSelector.addPropertyChangeListener("filename", propertyChangeListener);
```
There are so many code formatting changes in this file, I really have problems identifying your code changes in these parts! Please revert all your code formatting changes, so your PR only contains the changed code. If you want to change the code formatting style, we can discuss it and do it in a separate commit and PR.

McPringle commented on src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java:
```diff
@@ -666,6 +668,54 @@ protected Type configure(final JFileChooser outputFileChooser) {
+
+		tileCachePathSelector.addPropertyChangeListener("filename", propertyChangeListener);
+
+		final JLabel lblTileCacheTimeLimit = new JLabel("Tile cache age limit");
```
Tile cache *time* limit?

McPringle commented on src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java:
```diff
@@ -745,6 +795,8 @@ public void setConfiguration(final Configuration c) {
 		totalTimeSpinner.setValue(c.getTotalTime());
 		backgroundMapVisibilitySlider.setValue((int) (c.getBackgroundMapVisibility() * 100));
 		photoTimeSpinner.setValue(c.getPhotoTime());
+		tileCacheTimeLimitSpinner.setValue(1000L*c.getTileCacheTimeLimit());
```
Seconds to milliseconds? A short hint as a comment would be nice.

n76 commented:
I think I've addressed your concerns. I made some other changes to the tile caching object that tie in indirectly. First I always use the uncached download method so the user agent, etc. are all in one place and used consistently. Secondly, I try to treat a cache failure by falling back to uncached behavior. Finally, I decided to add a file extension to the cached tiles so they can easily be opened by image handling apps by double clicking on them.

Looking at my cache directory, I think I need to handle the aging differently. At present I only age the files we are looking at. This means that if/when you finish making a video the cached tiles are likely to remain for a long time (until you make another animation that happens to use those same tiles).

McPringle commented:
This PR closes issue #77.
